File: frontend/src/types/comparison.ts
export interface AddedComponent {
  component_identifier: string | null
  component_name: string | null
  component_ticker: string | null
  weight_new: number | null
  shares_new: number | null
  component_sector: string | null
  component_currency: string | null
}

export interface RemovedComponent {
  component_identifier: string | null
  component_name: string | null
  component_ticker: string | null
  weight_old: number | null
  shares_old: number | null
  component_sector: string | null
  component_currency: string | null
}

export interface WeightChange {
  component_identifier: string | null
  component_name: string | null
  component_ticker: string | null
  weight_old: number | null
  weight_new: number | null
  weight_delta: number | null
  component_sector: string | null
  shares_old: number | null
  shares_new: number | null
  shares_delta: number | null
}

export interface CompareResponse {
  ticker: string
  date1: string
  date2: string
  summary: {
    added_count: number
    removed_count: number
    weight_changes_count: number
  }
  added: AddedComponent[]
  removed: RemovedComponent[]
  weight_changes: WeightChange[]
}

export interface ChangeEvent {
  date_from: string
  date_to: string
  change_type: string
  component_identifier: string | null
  component_name: string | null
  component_ticker: string | null
  component_sector: string | null
  weight_old: number | null
  weight_new: number | null
  weight_delta: number | null
}

export interface ChangesResponse {
  ticker: string
  changes: ChangeEvent[]
}

File: backend/schemas/comparison.py
from pydantic import BaseModel
from typing import Optional


class AddedComponent(BaseModel):
    component_identifier: Optional[str] = None
    component_name: Optional[str] = None
    component_ticker: Optional[str] = None
    weight_new: Optional[float] = None
    component_sector: Optional[str] = None
    component_currency: Optional[str] = None


class RemovedComponent(BaseModel):
    component_identifier: Optional[str] = None
    component_name: Optional[str] = None
    component_ticker: Optional[str] = None
    weight_old: Optional[float] = None
    component_sector: Optional[str] = None
    component_currency: Optional[str] = None


class WeightChange(BaseModel):
    component_identifier: Optional[str] = None
    component_name: Optional[str] = None
    component_ticker: Optional[str] = None
    weight_old: Optional[float] = None
    weight_new: Optional[float] = None
    weight_delta: Optional[float] = None
    component_sector: Optional[str] = None
    shares_old: Optional[float] = None
    shares_new: Optional[float] = None
    shares_delta: Optional[float] = None


class CompareSummary(BaseModel):
    added_count: int
    removed_count: int
    weight_changes_count: int


class CompareResponse(BaseModel):
    ticker: str
    date1: str
    date2: str
    summary: CompareSummary
    added: list[AddedComponent]
    removed: list[RemovedComponent]
    weight_changes: list[WeightChange]


class ChangeEvent(BaseModel):
    date_from: str
    date_to: str
    change_type: str
    component_identifier: Optional[str] = None
    component_name: Optional[str] = None
    component_ticker: Optional[str] = None
    component_sector: Optional[str] = None
    weight_old: Optional[float] = None
    weight_new: Optional[float] = None
    weight_delta: Optional[float] = None


class ChangesResponse(BaseModel):
    ticker: str
    changes: list[ChangeEvent]
